### ISSUE TITLE
e2e flake: deployment should label adopted RSs and pods [minUnavailable/maxSurge violated]

### ISSUE THREAD

**Issue by janetkuo** [kubernetes/kubernetes#21809]
Looks similar to #19299 

https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/21795/kubernetes-pull-build-test-e2e-gce/30065/?debugUI=CLOUD
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:72
  deployment should label adopted RSs and pods [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:71

  Expected error:
      <*errors.errorString | 0xc20854c6d0>: {
          s: "total pods available: 0, less than the min required: 2",
      }
      total pods available: 0, less than the min required: 2
  not to have occurred

  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:849
```

The pod-template-hash label are added as expected; there's a new RS with 3 replicas as expected; however, all 3 the pods are unavailable. Those 3 pods are adopted and should be ready before the deployment is created. Looks like those pods are killed and created? Looking at the event, 6 pods are created (but should be only 3). 

```
Feb 23 14:33:08.649: INFO: event for nginx-controller-3j1uw: {default-scheduler } Scheduled: Successfully assigned nginx-controller-3j1uw to e2e-gce-builder-2-2-minion-n1u1
Feb 23 14:33:08.649: INFO: event for nginx-controller-3j1uw: {kubelet e2e-gce-builder-2-2-minion-n1u1} Pulling: pulling image "nginx"
Feb 23 14:33:08.649: INFO: event for nginx-controller-3j1uw: {kubelet e2e-gce-builder-2-2-minion-n1u1} Pulled: Successfully pulled image "nginx"
Feb 23 14:33:08.649: INFO: event for nginx-controller-3j1uw: {kubelet e2e-gce-builder-2-2-minion-n1u1} Created: Created container with docker id e0b3b69e8aa5
Feb 23 14:33:08.649: INFO: event for nginx-controller-3j1uw: {kubelet e2e-gce-builder-2-2-minion-n1u1} Started: Started container with docker id e0b3b69e8aa5
Feb 23 14:33:08.649: INFO: event for nginx-controller-8vc1j: {default-scheduler } Scheduled: Successfully assigned nginx-controller-8vc1j to e2e-gce-builder-2-2-minion-zi4d
Feb 23 14:33:08.649: INFO: event for nginx-controller-8vc1j: {kubelet e2e-gce-builder-2-2-minion-zi4d} Pulling: pulling image "nginx"
Feb 23 14:33:08.649: INFO: event for nginx-controller-8vc1j: {kubelet e2e-gce-builder-2-2-minion-zi4d} Pulled: Successfully pulled image "nginx"
Feb 23 14:33:08.649: INFO: event for nginx-controller-8vc1j: {kubelet e2e-gce-builder-2-2-minion-zi4d} Created: Created container with docker id b57072796d7d
Feb 23 14:33:08.649: INFO: event for nginx-controller-8vc1j: {kubelet e2e-gce-builder-2-2-minion-zi4d} Started: Started container with docker id b57072796d7d
Feb 23 14:33:08.649: INFO: event for nginx-controller-900jh: {default-scheduler } Scheduled: Successfully assigned nginx-controller-900jh to e2e-gce-builder-2-2-minion-zi4d
Feb 23 14:33:08.649: INFO: event for nginx-controller-h9bp9: {default-scheduler } Scheduled: Successfully assigned nginx-controller-h9bp9 to e2e-gce-builder-2-2-minion-a4y3
Feb 23 14:33:08.649: INFO: event for nginx-controller-onb6n: {default-scheduler } Scheduled: Successfully assigned nginx-controller-onb6n to e2e-gce-builder-2-2-minion-a7mz
Feb 23 14:33:08.649: INFO: event for nginx-controller-onb6n: {kubelet e2e-gce-builder-2-2-minion-a7mz} Pulling: pulling image "nginx"
Feb 23 14:33:08.649: INFO: event for nginx-controller-onb6n: {kubelet e2e-gce-builder-2-2-minion-a7mz} Pulled: Successfully pulled image "nginx"
Feb 23 14:33:08.649: INFO: event for nginx-controller-onb6n: {kubelet e2e-gce-builder-2-2-minion-a7mz} Created: Created container with docker id 750ac3422eab
Feb 23 14:33:08.649: INFO: event for nginx-controller-onb6n: {kubelet e2e-gce-builder-2-2-minion-a7mz} Started: Started container with docker id 750ac3422eab
Feb 23 14:33:08.649: INFO: event for nginx-controller-pgkjb: {default-scheduler } Scheduled: Successfully assigned nginx-controller-pgkjb to e2e-gce-builder-2-2-minion-1pst
Feb 23 14:33:08.649: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-8vc1j
Feb 23 14:33:08.649: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-3j1uw
Feb 23 14:33:08.649: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-onb6n
Feb 23 14:33:08.649: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-h9bp9
Feb 23 14:33:08.649: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-900jh
Feb 23 14:33:08.649: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-pgkjb
```

cc @bgrant0607 @kargakis @mqliang @nikhiljindal @ironcladlou @soltysh

**Comment by bgrant0607** [comment]
cc @bprashanth @nikhiljindal

**Comment by AdoHe** [comment]
Another occurance:
```
Deployment
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:72
  deployment should label adopted RSs and pods [It]
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:71

  Test Panicked
  runtime error: invalid memory address or nil pointer dereference
  /usr/src/go/src/runtime/panic.go:387

  Full Stack Trace
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/deployment/deployment.go:194 (0xa147be)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/deployment/deployment.go:157 (0xa140ca)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/deployment/deployment.go:73 (0xa12bd9)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/deployment/deployment.go:58 (0xa12a10)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/util.go:2111 (0x56d90f)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:181 (0x8ca750)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:143 (0x8ca56b)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:137 (0x8ca4bc)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/util.go:2148 (0x531d20)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:857 (0x490319)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:70 (0x58cc6a)
  /go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/onsi/ginkgo/internal/leafnodes/runner.go:109 (0x13b05f5)
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e_test.go:28 (0x542118)
  /usr/src/go/src/testing/testing.go:447 (0x46112f)
  /usr/src/go/src/runtime/asm_amd64.s:2232 (0x440541)
```
[https://console.cloud.google.com/m/cloudstorage/b/kubernetes-jenkins/o/pr-logs/pull/21535/kubernetes-pull-build-test-e2e-gce/30139/build-log.txt](https://console.cloud.google.com/m/cloudstorage/b/kubernetes-jenkins/o/pr-logs/pull/21535/kubernetes-pull-build-test-e2e-gce/30139/build-log.txt)
seems #21795 can fix this.

**Comment by janetkuo** [comment]
When the `total pods available: 0, less than the min required: 2` failure happens, new pods are created by the adopted RS instead of just labeling existing pods. This flake is observed pretty often in #21795. Not sure if it happens elsewhere. 

https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/pr-logs/pull/21795/kubernetes-pull-build-test-e2e-gce/30230/artifacts/?debugUI=CLOUD

**Comment by quinton-hoole** [comment]
I saw this test fail in http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-ubernetes-lite/384/ at Feb 23, 2016 8:23:22 PM.

```

20:46:57 • Failure [60.751 seconds]
20:46:57 Deployment
20:46:57 /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:72
20:46:57   deployment should label adopted RSs and pods [It]
20:46:57   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:71
20:46:57 
20:46:57   Expected
20:46:57       <int>: 0
20:46:57   to be >
20:46:57       <int>: 0
20:46:57 
20:46:57   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:883
```

**Comment by janetkuo** [comment]
The above failure:
https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-ubernetes-lite/384/?debugUI=CLOUD
```
20:46:57 Feb 23 20:46:40.539: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-j0pzb
20:46:57 Feb 23 20:46:40.539: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-baoun
20:46:57 Feb 23 20:46:40.539: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-02e77
20:46:57 Feb 23 20:46:40.539: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-ikmbt
20:46:57 Feb 23 20:46:40.539: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-b7hv8
20:46:57 Feb 23 20:46:40.539: INFO: event for nginx-controller: {replicaset-controller } SuccessfulCreate: Created pod: nginx-controller-9t5en
20:46:57 Feb 23 20:46:40.539: INFO: event for nginx-controller: {replicaset-controller } SuccessfulDelete: Deleted pod: nginx-controller-9t5en
20:46:57 Feb 23 20:46:40.539: INFO: event for nginx-controller: {replicaset-controller } SuccessfulDelete: Deleted pod: nginx-controller-ikmbt
```

There were already 3 ready pods. Then the adopted RS created 3 more, and at least one of the pods the deployment manages doesn't contain pod-template-hash label.

**Comment by bprashanth** [comment]
Is it possible to surface RS overshoots in the error message (or clarify the 0 != 0 message)?

**Comment by janetkuo** [comment]
or, it's actually 0 !> 0 message

**Comment by janetkuo** [comment]
The first failure mode is due to hash value changing after labeling pod template and should be solved by #21795.

**Comment by bgrant0607** [comment]
This appears to be fixed, and the test was not moved to the Flaky suite, so closing.
